testng-core/src/main/java/org/testng/internal/Graph.java

```diff
@@ -174,6 +175,12 @@ private static void ppp(String s) {
     }
   }
 
+  private static void ppp(Supplier<String> s) {
```

Comment:
Fyi, ppp methods should be replaced by loggers https://github.com/cbeust/testng/issues/2589

Reply:
That should better be addressed later. I tried throwing exception from `Test#toString` in order to remove all (?) `toString` calls by the framework, and `toString` inside `ppp` was easy to avoid by adding `Supplier`.